docs/proposals/2017-04-27 Text Formatting.md

```diff
@@ -0,0 +1,239 @@
+
+This is useful for:
+
+* _Importing Content_ - Many legacy CMSes, including ones in use at Washington Post, store content as HTML, and thus having a direct pathway for sections of HTML content into ANS is expedient;
+* _WYSIWYG Web Editor_ - Virtually all web browsers support the "contenteditable" attribute to some degree, so using HTML for some of the formatting allows web based editors (such as Ellipsis) to take advantage of that fact;
+* _HTML Templates_ - Simple HTML templates can copy the text value as-is with no additional parsing or rendering.
+
+Any successor solution should seek to be as useful or more to the above end-users.
+
+
+## Goals of this Proposal
+
+According to this proposal, the ideal text formatting solution should also be feasible to use in:
+
+* _Rainbow_ - Existing rainbow products already consume JSON formatted content;
+* _iOS, Android, Apple News, and other non-HTML environments_ - Apple in particular is enthusiastic about its design-oriented Apple News format, which has features without direct HTML analogues;
+* _Elasticsearch_ - It is a critical piece of the Content API with unique limitations on how data is stored that should be addressed in the schema when possible (HTML is not the worst, not the best format here).
+
+It should also:
+
+* Maintain existing HTML options (when practical) as a convenience options for current & similar ANS users;
+* Limit text formatting to a reasonable set of options (somewhere in the range of "superscript yes, kerning no");
+* Address existing limitations in representing element-element relationships that are not strictly sequential, and element-document relationships.
+
+
+# Proposal
+
+## Part I. Structured Text Object.
+
+### `attributed_text` object describes structured text content
+
+Attributed text object will have the following structure. Note that proposed structure is compatible with preexisting `text` story element (as of 0.5.8).
+
+* a required `text` property with plain text content of type `string`,
+* an optional `attributes` property with array of `attribute` objects describing formatting to be applied to the text,
+* an optional `selections` property with array of `selection` objects describing text sub-strings and formatting,
+* an optional `content` property with HTML formatted text content.
+
+Example:
+```
+  {
+    "_id": "AAAAAAAAAAAAAAAAAAAAAAAAAA",
+    "type": "text",
+    "text": "The text attribute should contain plain text content only.",
+    "selections": [ ... ]
+    "content": "The <b>text</b> attribute should contain <i>plain text</i> content only."
+  }
+```
+
+### `selection`s describe formatting for specific segments of a text string
+
+Each `selection` object defines a part of a text string and what formatting attributes should be applied to it. An idea borrowed from Apple and Twitter.
+
+* a required `start` and `length` integer properties define sub-string range,
+* an optional `attributes` property with array of `attribute` objects.
```

Comment:
If I have text that I need to show up in an editor app but I don't want displayed in a published story, such as a note between a copy editor and a reporter, would there be a way to denote that with this schema?

Reply:
My bias is that content that isn't directly or indirectly reader-facing should not be included in ANS. But you could certainly save+update a "notes" document that keys off of a (story id + revision id + content element id) and is loaded in an editor alongside a story revision

Reply:
That's my feeling as well. This kind of notes would be very similar to "annotations", and this proposal is really about formatting. Those should live in separate places. Though, I would agree that the technical requirements would be very similar. There is still the concept of a text selection. Maybe we need to make a separate proposal to address those use cases.

Reply:
I agree with the distinction between reader-facing vs private/internal, but I do think that this spec should be updated to support annotations which _are_ reader-facing but not specifically 'formatting' (as in bold, underline, italic, etc).

For example, we've discussed the new 'notes' story tools feature, which allows a popup document to be displayed when they reader hovers or taps on a certain selection of text. This is sort of similar to the 'Genius' notes system that The Fix was using last year.  My preference is to have something like this included in the `attributes` along with other formatting and presentation attributes.  

Another good reason to separate internal/private annotations is that we probably want to avoid those being sent to user-facing clients in the JSON document, even if the client doesn't render them. Exposing something like editor notes even in the JSON API could potentially lead to an embarrassing situation.

So for me, the distinction is really more of an access/visibility concern, rather than the type of attribute/annotation.

Reply:
Though, I should add that this also gets to the point @calexicoz brought up above, in the case that these annotations are being added by a separate tool from the original editor.  This, though, I think is honestly more of a problem that needs to be solved on the content API side as opposed to the schema.